obsproject/obs-studio
# linux-v4l2: Save device by id or path

### Description

Use /dev/v4l/by-id or /dev/v4l/by-path when available in order to persist device binding to specific source.
Fixes #3003
Closes #3437

### Motivation and Context

Currently v4l2 uses /dev/videoX which is not consistent across reboots and device reconnections.

This patch uses directories:

* /dev/v4l/by-id - User has 2 devices: C920 and microscope. Sometimes I use them with another computer. I want to keep them assigned to specific sources/scenes without remembering which USB port to use.
* /dev/v4l/by-path - User has 2 devices: front camera connected to first usb and top camera connected to second usb. If one of the devices is upgraded/changed it remains assigned to the same path.
* /dev/videoX - Devices which are not present in above directories (eg. loopback)

The device can exists twice on the list (by-id vs by-path addressing). It's up to user which connect to the source.

In case of by-path device the source name consist of "bus_info (device name)"
Otherwise it uses current naming: "device name (bus_info)"

Fixes #3003
Closes #3437

### How Has This Been Tested?

Tested with 5 different v4l2 devices:

ID 04fc:6353 Sunplus Technology Co., Ltd General USB Device (Lamax Electra)
ID 0c45:6366 Microdia Webcam Vitade AF (Digital microscope)
ID 534d:2109 MacroSilicon USB3.0 HD VIDEO (Video grabber)
ID 046d:082d Logitech, Inc. HD Pro Webcam C920
ID 046d:085e Logitech, Inc. BRIO Ultra HD Webcam

Test Case 1:

- Plug all the devices, assign them by id, put them on the scene
- It should display all the feeds

Test Case 2:

- Having all devices assigned by id switch their USB ports.
- The scene shouldn't change, all feeds in their old places.

Test Case 3:

- Having all devices assigned by id turn off OBS
- Run OBS
- Nothing is on the scene
- Plug devices in another order
- Devices should be assigned to correct places by udev

Test Case 4:

- Plug 2 devices, assign them by path, put them on the scene
- Switch USB ports
- The scene should have switched cameras

Test Case 5:

- Create v4l2loopback: `sudo modprobe v4l2loopback devices=1 video_nr=88 ; ffmpeg -re -f lavfi -i "movie=something.mp4" -f v4l2 /dev/video88`
- The v4l2loopback should be available on the list and working

### Types of changes
- Tweak (non-breaking change to improve existing functionality)

### Checklist:
